src/content/reference/rules/components-and-hooks-must-be-pure.md

```diff
@@ -0,0 +1,283 @@
+</Intro>
+
+<InlineToc />
+
+<DeepDive>
+#### Why does render need to be pure? {/*why-does-render-need-to-be-pure*/}
+
+React is declarative: you tell React what to render in your component's logic, and React will figure out how best to display it to your user.
+
+When render is kept pure, React can understand how to prioritize which updates are most important for the user to see first. This is made possible because of render purity: since components don't have side effects in render, React can pause rendering components that aren't as important to update, and only come back to them later when it's needed.
+
+Concretely, this means that rendering logic can be run multiple times in a way that allows React to give your user a pleasant user experience. However, if your component has an untracked side effect – like modifying the value of a global variable during render – when React runs your rendering code again, your side effects will be triggered in a way that won't match what you want. This often leads to unexpected bugs that can degrade how your users experience your app.
+</DeepDive>
+
+---
+
+## Components must be idempotent {/*components-must-be-idempotent*/}
+
+Components must always return the same output with respect to their inputs – props, state, and context. This is known as _idempotency_.
+
+[Idempotency](https://en.wikipedia.org/wiki/Idempotence) is a term popularized in functional programming. It refers to the idea that you [always get the same result everytime](learn/keeping-components-pure) you run that piece of code with the same inputs.
+
+```js {3}
+function NewsFeed({ items }) {
+  // ✅ Array.filter is idempotent: it doesn't mutate `items`
+  const filteredItems = items.filter(item => item.isDisplayed === true);
+  return (
+    <ul>
+      {filteredItems.map(item => <li key={item.id}>{item.text}</li>)}
+    </ul>
+  );
+}
+```
+
+This means that _all_ code that runs during render must also be idempotent in order for this rule to hold. For example, this line of code is not idempotent (and therefore, neither is the component):
+
+```js {2}
+function Clock() {
+  const date = new Date(); // ❌ always returns a different result!
+  return <div>{date}</div>
+}
+```
+
+`new Date()` is not idempotent as it always returns the current date and changes its result every time it's called. When you render the above component, the time displayed on the screen will stay stuck on the time that the component was rendered. Similarly, functions like `Math.random()` also aren't idempotent, because they return different results every time they're called, even when the inputs are the same.
+
+This doesn't mean you shouldn't use non-idempotent functions like `new Date()` _at all_ – you should just avoid using them during render. One quick heuristic to tell if code runs during render is to examine where it is: if it's written at the top level like in the example above, there's a good chance it runs during render.
```

Comment:
Can we provide an example of how we can use `new Date()` outside of render? I get a little bit struck at this part. May be adding example should provide better understanding?

Reply:
Sure, added one!